wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,310 @@
+        let mut headers = http::HeaderMap::new();
+        headers.insert(
+            http::header::CONTENT_TYPE,
+            HeaderValue::from_static(CONTENT_TYPE_FORM),
+        );
+
+        let body = serde_urlencoded::to_string([
+            ["log", self.credentials.username.as_str()],
+            ["pwd", self.credentials.password.as_str()],
+            ["rememberme", "true"],
+            [
+                "redirect_to",
+                self.api_base_url.derived_rest_nonce_url().as_str(),
+            ],
+        ])
+        .ok()?;
+
+        Some(WpNetworkRequest {
+            method: RequestMethod::POST,
+            url: self.api_base_url.derived_wp_login_url().into(),
+            header_map: WpNetworkHeaderMap::new(headers).into(),
+            body: Some(WpNetworkRequestBody::new(body.into_bytes()).into()),
+        })
+    }
+
+    async fn nonce_from_request(&self, request: WpNetworkRequest) -> Option<String> {
+        self.request_executor
+            .execute(request.into())
+            .await
+            .ok()
+            .and_then(|response| {
+                // A 200 OK response from the `rest_nonce_url` (a.k.a `wp-admin/admin-ajax.php`)
+                // should be the nonce value. However, just in case the site is configured to
+                // return a 200 OK response with other content (for example redirection to
+                // other webpage), here we check the body length here for a light validation of
+                // the nonce value.
+                if response.status_code == 200 {
+                    let body = response.body_as_string();
+                    if body.len() < 50 {
+                        return Some(body);
+                    }
+                }
+                None
+            })
+    }
+}
+
+#[async_trait::async_trait]
+impl Authenticator for CookieAuthenticator {
+    fn host(&self) -> &str {
+        self.api_base_url.host_str()
+    }
+
+    fn should_authenticate(&self, request_url: &str, response_status_code: Option<u16>) -> bool {
+        let is_same_host = Url::parse(request_url)
+            .ok()
+            .map(|f| f.host_str() == Some(self.host()));
+        if is_same_host != Some(true) {
+            return false;
+        }
+
+        if let Some(response_status_code) = response_status_code {
+            return response_status_code == 401;
+        }
+
+        true
+    }
+
+    async fn authentication_headers(&self) -> Option<HeaderMap> {
+        self.get_rest_nonce().await.map(|nonce| {
+            let mut headers = HeaderMap::new();
+            headers.insert(
+                "X-WP-Nonce",
+                nonce
+                    .try_into()
+                    .expect("This conversion should never fail since nonce is a short string"),
+            );
+            headers
+        })
+    }
+
+    fn reset(&self) {
+        *self.nonce.write().expect("Failed to unlock nonce") = None;
```

Comment:
This `.expect` looks wrong to me. There are valid cases where the lock wouldn't be released to us. On the contrary, this looks like a case we need to handle with a `Result` and maybe retry after a grace period.

Reply:
I can turn this into an if-let and only handle the ok case, which should be safer.

> There are valid cases where the lock wouldn't be released to us.

Can you expand on this?

> **Poisoning**
>
> An RwLock, like [Mutex](https://doc.rust-lang.org/std/sync/struct.Mutex.html), will become poisoned on a panic. Note, however, that an RwLock may only be poisoned if a panic occurs while it is locked exclusively (write mode). If a panic occurs in any reader, then the lock will not be poisoned.

Reading the api doc, it doesn't look like there is much we can do about the error.

Reply:
Hmm.. I completely misremembered how `RwLock` works. `read` & `write` blocks the current thread until the lock is available, instead of returning an error as `try_read` & `try_write` does. 🤔 Sorry about that.

In that case, the lock _could_ be used to address the consecutive requests issue by retrieving the lock to `write` when the authentication process is started. Then any `read` access will be blocked until the authentication is complete.

We probably need to keep track of our authentication attempts, otherwise we'll keep trying to authenticate even if we have bad credentials - but that doesn't have to happen in this PR.